denoise_tests.py

```diff
@@ -0,0 +1,385 @@
+import argparse
+import logging
+import os
+import sys
+from timeit import default_timer as timer
+
+import cv2 as cv
+import bm3d
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import psutil
+import skimage
+import tifffile
+import zarr
+from scipy import fftpack
+from skimage.exposure import rescale_intensity
+from skimage.restoration import (
+    estimate_sigma,
+    denoise_nl_means,
+    denoise_bilateral,
+    denoise_tv_chambolle,
+    denoise_wavelet
+)
+
+import compress_zarr
+
+try:
+    import imagej
+    import scyjava
+    # initialize Fiji with the external non-local means plugin
+    IMAGEJ = imagej.init(['sc.fiji:fiji', 'com.github.thorstenwagner:ij-nl-means'], headless=False)
+    IMAGEJ.ui().showUI()
+except ImportError:
+    IMAGEJ = None
+
+
+def identity(data):
+    return data
+
+
+def run_ij_plugin(arr, plugin, args):
+    """Running external ImageJ plugins does not work headlessly,
+    for debug use only."""
+    # Set the active image in the Window Manager
+    IMAGEJ.ui().show(IMAGEJ.py.to_java(arr))
+    # Run plugin on the active image
+    IMAGEJ.py.run_plugin(plugin, args)
+    # Get active image
+    imp = IMAGEJ.py.active_image_plus()
+    # Convert to numpy ndarray
+    result = IMAGEJ.py.from_java(imp)
+    # Hack to bypass the "Save image before closing?" prompt
+    imp.changes = False
+    imp.close()
+    return result
+
+
+# OpenCV Total variation denoising
+# This only works with 8-bit images
+def cv_tv(data):
+    data = skimage.img_as_ubyte(data)
+    result = np.zeros_like(data)
+    if data.ndim == 2:
+        cv.denoise_TVL1([data], result)
+    elif data.ndim == 3:
+        for i in range(data.shape[0]):
+            cv.denoise_TVL1([data[i]], result[i])
+    else:
+        raise ValueError
+    return result
+
+
+# OpenCV Block-matching and 3D filtering
+# Only available in non-free extension
+# I was unable to compile it so this remains untested
+def cv_bm3d(data):
+    result = np.zeros_like(data)
+    if data.ndim == 2:
+        cv.xphoto.bm3dDenoising(data, result)
+    elif data.ndim == 3:
+        for i in range(data.shape[0]):
+            cv.xphoto.bm3dDenoising(data[i], result[i])
+    else:
+        raise ValueError
+    return result
+
+
+# OpenCV non-local means denoising
+# Much faster than the scikit-image version
+def cv_nl_means(data, h):
+    norm_type = cv.NORM_L1  # required for uint16
+    if data.ndim == 2:
+        return cv.fastNlMeansDenoising(data, h=h, normType=norm_type)
+    elif data.ndim == 3:
+        slices = []
+        for s in data:
```

Comment:
Looks like this is a 2D-only method.  I would be nervous about something that depends on the global mean of the slice (which I think this does?) creating inconsistencies in Z.

Reply:
Agreed, as far as I could tell all of the OpenCV methods are designed for 2D or multichannel images